casanovo/denovo/model.py

```diff
@@ -176,12 +193,592 @@ def forward(
+                if "" not in aa_list:
+                    # Remove stop token as needed, i.e. if mass filter induced termination
+                    if "$" in aa_list:
+                        aa_list.remove("$")
+
+                    pep_str = "".join(aa_list)
+                    pred_mz = self.peptide_mass_calculator.mass(
+                        aa_list, charge=precursors[i, 1].item()
+                    )
+                    delta_mass_ppm = (pred_mz - prec_mz) / prec_mz * 10**6
+                    isPrecursorFit = (
+                        abs(delta_mass_ppm) <= self.precursor_mass_tol
+                    )
+
+                    # Check if predicted seq already in cache
+                    if pep_str not in cache_seq_dict[spec_idx]:
+                        # Directly cache if we don't already have k peptides cached
+                        if insert_idx < (spec_idx + 1) * beam:
+                            cache_tokens[insert_idx, :] = tokens[i, :]
+                            cache_scores[insert_idx, :, :] = scores[i, :, :]
+                            cache_idx_dict[spec_idx] += 1  # move the pointer
+                            cache_seq_dict[spec_idx].add(pep_str)
+                            smx = self.softmax(scores)[i, :, :]
+
+                            aa_scores = [
+                                smx[idx][self.decoder._aa2idx[aa]].item()
+                                for idx, aa in enumerate(aa_list)
+                            ]
+                            pep_score = np.mean(aa_scores)
+
+                            # Cache peptides with fitting and non-fitting precursor mass separately
+                            if isPrecursorFit:
+                                heappush(
+                                    cache_score_dict[spec_idx][0],
+                                    (pep_score, insert_idx),
+                                )
+                            else:
+                                heappush(
+                                    cache_score_dict[spec_idx][1],
+                                    (pep_score, insert_idx),
+                                )
+
+                        else:
+                            smx = self.softmax(scores)[i, :, :]
+                            aa_scores = [
+                                smx[idx][self.decoder._aa2idx[aa]].item()
+                                for idx, aa in enumerate(aa_list)
+                            ]
+                            new_score = np.mean(aa_scores)
+
+                            # Cache fitting peptide either if non-fitting cached
+                            # or higher confidence than lowest scoring fitting peptide cached
+                            if isPrecursorFit:
+                                # Check if any non-fitting peptide cached
+                                if len(cache_score_dict[spec_idx][1]) > 0:
+                                    (
+                                        pop_pep_score,
```

Comment:
**todo:** No need to capture the results from the `heappop` because they're not used further.